sapicore/engine/ensemble/__init__.py

```diff
@@ -0,0 +1,66 @@
+""" Ensembles (layers) manage neuron collections by fixing their size, extending attribute tensors,
+and handling layer-aware operations such as heterogeneous parameter initialization
+and normalization.
+
+Note
+----
+Ensemble instance attributes are 1D tensors corresponding to potentially heterogeneous parameters
+and states of neurons (tensor elements) whose behavior is prescribed by their parent class
+(e.g., :class:`~engine.neuron.spiking.LIF.LIFNeuron`).
+
+Ensembles can be initialized programmatically or from a configuration YAML containing neuron parameters and their
+generation method (fixed, zipped, grid, or drawn from a distribution). Forward calls invoke the parent neuron class
+method.
+
+Example
+-------
+Initialize an ensemble of basic analog neurons, inheriting default parameters and behavior from
+:class:`~engine.neuron.analog.AnalogNeuron`:
+
+    >>> from sapicore.engine.ensemble.spiking import LIFEnsemble
+    >>> layer = LIFEnsemble(num_units = 10)
+
+"""
+import torch
+from torch import Tensor
+
+from sapicore.engine.neuron import Neuron
+
+
+class Ensemble(Neuron):
+    """Ensemble base class. Provides generic implementations of tensor expansion and parameter diversification."""
+
+    def __init__(self, num_units: int = 1, **kwargs):
+        """Constructs a generic ensemble instance, inheriting attributes from :class:`~engine.neuron.Neuron`."""
+        super().__init__(**kwargs)
+        self.num_units = num_units
+
+        # expand configurable parameter tensors to size `num_units`.
+        for prop in self._config_props_:
+            temp = getattr(self, prop)
+            setattr(self, prop, torch.zeros(self.num_units, dtype=torch.float, device=self.device) + temp)
+
```

Comment:
It's a little scary to blindly assume all config properties should be duplicated to `num_units`. That almost def won't be true of user classes. You either need to curate this list manually or have a class function that returns this list of props that should be blown up to num units or some other way of tracking props that need that.

Reply:
There's also an assumption there that neuron classes, when combined with ensemble will work as intended. But a neuron class author may not remember that. So it's even more important that neuron class authors be aware that the class may be used with an ensemble by having the num_units be defined at the neuron level - even if it's set to 1.

Ideally whoever writes the neuron would write the ensemble using that neuron so they can make sure it works. So writing this clearly in the docs as part of `num_units` in neuron class may be helpful.

So, I think if a neuron class can explicitly list which props are safe to be duplicated by a ensemble subclass (e.g. `ensemble_properties: List[str] = [...]`) so that the subclass can do it safely using this list will solve both problems. Because if the neuron class author doesn't list any, it's assumed those variables must be the same for all units. But, for those they wrote the neuron base class with ensembles in mind, they can list it there to be used safely.

Reply:
Great points. I will add an attribute specifying what subset of `_config_props_` should be extended. It's conceivable that someone would want to include, e.g., a global scalar attribute in there. 

Generally, though, fields declared configurable for a network component would typically describe functional properties of the object being duplicated. Some scalars or strings, like `identifier`, are technically "configurable", but including them in `_config_props_` would be silly. Maybe I should add a docstring note to that effect.

Fantastic comments all around, as always. Thanks for taking the time!

Reply:
Added an `_extensible_props_` class attribute to Ensemble, with a setter that returns the full `_config_props` if None. Still assuming that all `_loggable_props_` should be extended (those would be voltage, input, spiked, etc.), but open to changing that.